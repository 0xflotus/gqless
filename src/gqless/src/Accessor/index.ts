export * from './Accessor'
export * from './FieldAccessor'
export * from './IndexAccessor'
export * from './RootAccessor'
