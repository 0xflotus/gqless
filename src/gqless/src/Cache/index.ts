export * from './Cache'
export * from './Value'
export * from './Ref'
