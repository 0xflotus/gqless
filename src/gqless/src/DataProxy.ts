import {
  ArrayNode,
  InnerNode,
  Node,
  NodeDataType,
  FieldsNode,
  FieldNode,
  Arguments,
  NodeContainer,
  ScalarNode,
} from './Node'
import { Extension, IExtension } from './Extension'

type RequiredKeys<T> = {
  [K in keyof T]-?: {} extends { [P in K]: T[K] } ? never : K
}[keyof T]

type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

type NodeExtensionType<
  TExtensions extends Record<string, Extension>,
  TNode extends Node<any>
> = TNode extends FieldsNode<any, any, infer Typename>
  ? TExtensions[Typename]
  : never

type NodeExtension<
  TExtensions extends Record<string, Extension>,
  TNode extends Node<any>
> = NodeExtensionType<TExtensions, TNode> extends Function
  ? ReturnType<NodeExtensionType<TExtensions, TNode>>
  : NodeExtensionType<TExtensions, TNode>

// Extracts the nullable value from a NodeContainer
type NodeContainerNullable<
  TNode extends Node<any>
> = TNode extends NodeContainer<any, infer TNullable, any>
  ? TNullable extends true
    ? null
    : never
  : never

// Gets the inner value for a NodeContainer
type NodeContainerValue<
  TNode extends NodeContainer<any, any, any>,
  TExtensions extends Record<string, Extension>
> = DataProxy<InnerNode<TNode>, TExtensions>

// Field argument function
type FieldArgFn<
  TNode extends FieldNode<any, any, any>,
  TExtensions extends Record<string, Extension>
> = TNode extends FieldNode<any, Arguments<infer T, infer TInputs>, any>
  ? 0 extends (1 & T & TInputs) // check for any (when undefined / null is passed)
    ? never
    : (RequiredKeys<NodeDataType<Arguments<T, TInputs>>> extends never
        ? (
            args?: NodeDataType<Arguments<T, TInputs>>
          ) =>
            | NodeContainerValue<TNode, TExtensions>
            | NodeContainerNullable<TNode>
        : (
            args: NodeDataType<Arguments<T, TInputs>>
          ) =>
            | NodeContainerValue<TNode, TExtensions>
            | NodeContainerNullable<TNode>)
  : never

// Returns a field + callback + nullable
type FieldProxy<
  TNode extends FieldNode<any, any, any>,
  TExtensions extends Record<string, Extension>
> = FieldArgFn<TNode, TExtensions> extends never
  ? (NodeContainerValue<TNode, TExtensions> | NodeContainerNullable<TNode>) // no args
  : (NodeContainerNullable<TNode> extends never
      ? (InnerNode<TNode> extends ScalarNode // always need to call scalar nodes
          ? FieldArgFn<TNode, TExtensions>
          : (FieldArgFn<TNode, TExtensions> &
              NodeContainerValue<TNode, TExtensions>)) // args + non-nullable = don't need to call it
      : FieldArgFn<TNode, TExtensions>) // args + nullable = need to call it

// Recurse over a FieldsNode (ObjectNode / InterfaceNode)
type RecurseFields<
  TNode extends FieldsNode<any, any, any, any>,
  TExtensions extends Record<string, Extension>
> = {
  [K in keyof NodeDataType<TNode>]: FieldProxy<
    TNode['fields'][K],
    TExtensions
  > &
    NodeExtension<TExtensions, TNode>[K]
} &
  {
    [K in keyof Omit<
      NodeExtension<TExtensions, TNode>,
      keyof NodeDataType<TNode>
    >]: NodeExtension<TExtensions, TNode>[K]
  }

// Recurse over ArrayNode + add nullable
type RecurseArray<
  TNode extends ArrayNode<any, any>,
  TExtensions extends Record<string, Extension>
> = {
  [K in keyof NodeDataType<TNode>]: K extends number
    ? (DataProxy<TNode['ofNode'], TExtensions> | NodeContainerNullable<TNode>)
    : NodeDataType<TNode>[K]
}

// Input factory type for Node
export type DataProxy<
  TNode extends Node<any>,
  TExtensions = unknown
> = TNode extends ArrayNode<any, any>
  ? RecurseArray<TNode, TExtensions>
  : TNode extends FieldsNode<any, any, any, any>
  ? RecurseFields<TNode, TExtensions>
  : NodeDataType<TNode>
