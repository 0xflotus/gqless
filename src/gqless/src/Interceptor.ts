import { createEvent } from '@gqless/utils'

import { AccessorInterceptor, accessorInterceptors } from './Node/abstract/Outputable'

export class Interceptor {
  public listening = false

  public onAccessor = createEvent<AccessorInterceptor>()

  public start() {
    accessorInterceptors.add(this.onAccessor.emit)
    this.listening = true
  }

  public stop() {
    accessorInterceptors.delete(this.onAccessor.emit)
    this.listening = false
  }
}
