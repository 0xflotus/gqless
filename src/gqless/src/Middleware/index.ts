export * from './defaultMiddleware'
export * from './Middleware'
export * from './MiddlewareEngine'
