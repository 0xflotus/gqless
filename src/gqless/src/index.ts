export * from './Node'
export * from './Cache'
export * from './Selection'
export * from './ASTBuilder'
export * from './GraphQL'
export * from './Plugin'
export * from './Interceptor'
export * from './Accessor'
export * from './Scheduler'
export * from './Type'
export * from './Variable'
export * from './helpers'
