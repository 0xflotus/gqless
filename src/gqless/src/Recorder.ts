import { Accessor } from './Accessor'
import { IAccessorRecorder, accessorRecorders } from './Node'
import { createEvent } from '@gqless/utils'

export class Recorder {
  public accessors = new Set<Accessor>()
  public isRecording = false

  private recordAccessor: IAccessorRecorder = accessor => {
    this.onAccessor.emit(accessor)
    this.accessors.add(accessor)
  }

  public onAccessor = createEvent<(accessor: Accessor) => void>()

  public start() {
    accessorRecorders.add(this.recordAccessor)
    this.isRecording = true
  }

  public stop() {
    accessorRecorders.delete(this.recordAccessor)
    this.isRecording = false
  }
}
