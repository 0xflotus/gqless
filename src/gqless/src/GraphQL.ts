import { DocumentNode } from 'graphql'
import { Selection, RootSelection } from './Selection'
import {
  ObjectNode,
  ArrayNode,
  FieldsNode,
  UnionNode,
  InterfaceNode,
} from './Node'
import { defaultMiddleware, MiddlewareEngine } from './Middleware'
import { ASTBuilder } from './ASTBuilder'
import { Scheduler } from './Scheduler'
import { Cache } from './Cache'
import {
  RootAccessor,
  Accessor,
  IndexAccessor,
  FieldAccessor,
} from './Accessor'
import { Disposable } from './mixins'

export type QueryResponse<Data = any> = { data: Data; errors: any }

export type QueryFetcher = (
  query: DocumentNode
) => Promise<QueryResponse> | QueryResponse

export type ProxyInterceptor = (
  target: unknown,
  prop: string | symbol
) => unknown

// @ts-ignore
export class GraphQL<TData = any> extends Disposable {
  public middleware = new MiddlewareEngine()
  public astBuilder = new ASTBuilder()
  public scheduler = new Scheduler(
    this.middleware,
    (selections, name) => this.fetchSelections(selections, name)!
  )
  public cache = new Cache()

  public selection = new RootSelection(this.node)
  public accessor = new RootAccessor(this.selection, this.cache, this.scheduler)

  public query: TData = this.accessor.data

  constructor(protected node: ObjectNode, protected fetchQuery: QueryFetcher) {
    super()

    this.middleware.add(...defaultMiddleware)

    this.selection.onSelect(selection => {
      this.middleware.all.onSelect(selection)
      this.scheduler.stage(selection)
    })

    this.selection.onUnselect(selection => {
      this.middleware.all.onUnselect(selection)
      this.scheduler.unstage(selection)
    })

    // this.selectionRoot.onSelectUpdate(selection => {
    //   this.middleware.all.onSelectUpdate(selection)
    //   this.batcher.stage(selection)
    // })
  }

  protected fetchSelections(selections: Selection<any>[], queryName?: string) {
    const result = this.astBuilder.buildDocument(queryName, ...selections)
    if (!result) return

    const responsePromise = (async () => {
      const response = await this.fetchQuery(result.doc)

      const recurseFieldsAccessor = (
        accessor: Accessor<Selection<ObjectNode | InterfaceNode | UnionNode>>,
        data: Record<string, unknown>
      ) => {
        if (data === null) return

        Object.entries(data).forEach(([dataProp, value]) => {
          // __typename is not a selection, but handled internally
          if (dataProp === '__typename') return

          let fieldAccessor = accessor.getChild(
            ({ selection }) => selection.dataProp === dataProp
          )

          if (!fieldAccessor) {
            const subSelection = accessor.selection.getField(
              s => s.dataProp === dataProp
            )

            if (!subSelection) {
              const dataPropMatch = dataProp.match(/^(.+)__(\d)$/)
              if (!dataPropMatch) {
                console.error(
                  `Couldn't locate selection ${accessor.path}.${dataProp}`
                )
                return
              }
              const [_, fieldName, alias] = dataPropMatch

              console.error(
                `Couldn't locate selection ${
                  accessor.path
                }.${fieldName} [alias=${alias}]`
              )

              return
            }

            fieldAccessor = new FieldAccessor(accessor, subSelection)
          }

          recurseAccessor(fieldAccessor, value)
        })
      }

      const recurseArrayAccessor = (
        accessor: Accessor<Selection<any>, IndexAccessor>,
        data: unknown[]
      ) => {
        if (data === null) return

        data.forEach((value, i) => {
          const indexAccessor =
            accessor.getChild(a => a.index === i) ||
            new IndexAccessor(accessor, i)

          recurseAccessor(indexAccessor, value)
        })
      }

      const recurseAccessor = (accessor: Accessor<any, any>, data: any) => {
        this.cache.update(accessor, data)

        if (accessor.node instanceof FieldsNode) {
          recurseFieldsAccessor(accessor, data)
        } else if (accessor.node instanceof ArrayNode) {
          recurseArrayAccessor(accessor, data)
        }
      }

      recurseAccessor(this.accessor, response!.data)

      return response
    })()

    this.middleware.all.onFetch(
      result.doc,
      responsePromise,
      queryName,
      selections
    )

    return responsePromise
  }

  public dispose() {
    super.dispose()
    this.scheduler.dispose()

    this.middleware.all.dispose()
  }
}
