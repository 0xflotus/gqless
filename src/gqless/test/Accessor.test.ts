import { RootAccessor, RootSelection, Scheduler, SelectionFetcher, Variable } from '../src'
import { Query } from './data'

jest.useFakeTimers()

const create = (fetchSelections: SelectionFetcher) => {
  const selection = new RootSelection(Query)
  const scheduler = new Scheduler(fetchSelections)
  const accessor = new RootAccessor(selection, scheduler)
  return { selection, scheduler, accessor }
}

const snapshotAccessor = (getMatches: (data: any) => void) => {
  const fetchSelections = jest.fn()
  const instance = create(fetchSelections)
  getMatches(instance.accessor.data)

  jest.runOnlyPendingTimers()

  expect(fetchSelections).toBeCalled()
  expect(instance.selection).toMatchSnapshot('selections')
  expect(instance.accessor).toMatchSnapshot('accessors')

  return instance
}

it('creates and schedules selections', () => {
  snapshotAccessor(data => {
    data.scalar
    data.object
    data.object.scalar
    data.arrayOfObjects.map((o: any) => o)
    data.arrayOfObjects[1].scalar
    data.arrayOfScalar
  })
})

it('supports arguments', () => {
  snapshotAccessor(data => {
    data.arrayOfObjects({ scalar: 2 })
    data.arrayOfObjects({ scalar: 3 })[0]
    data.arrayOfObjects({ scalar: 2 })[0].scalar
    data.arrayOfObjects({ input: new Variable({ scalar: 1 }) })
  })
})
