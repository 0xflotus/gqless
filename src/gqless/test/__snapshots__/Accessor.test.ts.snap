// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates and schedules selections: accessors 1`] = `
Query {
  scalar
  object {
    scalar
  }
  arrayOfObjects {
    0
    1 {
      scalar
    }
  }
  arrayOfScalar
}
`;

exports[`creates and schedules selections: selections 1`] = `
Query {
  scalar
  object {
    scalar
  }
  arrayOfObjects {
    scalar
  }
  arrayOfScalar
}
`;

exports[`supports arguments: accessors 1`] = `
Query {
  arrayOfObjects({"scalar":2}) {
    0 {
      scalar
    }
  }
  arrayOfObjects({"scalar":3}) {
    0
  }
  arrayOfObjects({"input":{"scalar":1}})
}
`;

exports[`supports arguments: selections 1`] = `
Query {
  arrayOfObjects({"scalar":2}) {
    scalar
  }
  arrayOfObjects({"scalar":3})
  arrayOfObjects({"input":{"scalar":1}})
}
`;
